### PR TITLE
planning: add open space algorithm documents

### PR DESCRIPTION
**Pull request by xiaoluauto**

### CODE REVIEW

File: docs/technical_documents/hybrid_a_star_en.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+Generate coarse trajectory on the open space
```
Comment (JasonZhou404) [inline docs/technical_documents/hybrid_a_star_en.md:1]: make it a title? Please follow the template

File: docs/technical_documents/open_space_trajectory_optimizer_en.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+Optimize trajectory based on the coarse trajectory.
```
Comment (JasonZhou404) [inline docs/technical_documents/open_space_trajectory_optimizer_en.md:1]: ditto

File: docs/technical_documents/open_space_trajectory_provider_en.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+Generate final trajectory based on the open space.
```
Comment (JasonZhou404) [inline docs/technical_documents/open_space_trajectory_provider_en.md:1]: ditto

File: docs/technical_documents/open_space_trajectory_provider_en.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    if (FLAGS_enable_open_space_planner_thread && !thread_init_flag_) {
+        task_future_ = cyber::Async(&OpenSpaceTrajectoryProvider::GenerateTrajectoryThread, this);
+        thread_init_flag_ = true;
+  }
+    ```
+4. Process3: whether the vehicle stop is due to fallback is determined by the IsVehicleStopDueToFallBack().This determines the final trajectory planning.
+5. Process4: if vehicle stop caused by fallback,replanning stitching trajectory by ComputeReinitStitchingTrajectory(). If not,replanning stitching trajectory by ComputeStitchingTrajectory(), please refer(https://github.com/ApolloAuto/apollo/blob/master/modules/planning/common/trajectory_stitcher.cc).
+6. Process5: The generation of trajectory depends on the FLAGS_enable_open_space_planner_thread. A stop trajectory is generated in the following cases: 1.planning thread is stop; 2.the vehicle arrives near the destination; 3. trajectory_error_ is triggered for more than 10 seconds; 4. previous frame planning failed. If the trajectory can be updated normally, the optimized trajectory is output normally. 
+
+7. Output: the optput is final trajectory information.
```
Comment (JasonZhou404) [inline docs/technical_documents/open_space_trajectory_provider_en.md:34]: typo

File: docs/technical_documents/open_space_trajectory_optimizer_en.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+9. Output: the optput is optimized trajectory information,
+
+
+# Algorithm Detail
+    ```cpp
+    LoadHybridAstarResultInEigen(&partition_trajectories[i], &xWS_vec[i],&uWS_vec[i])
+    ```
+    the function is to transform the initial trajectory information into the form needed for optimization.
+1. parameter:the initial trajectory and parameter matrix.
+2. introduction:the trajectory information is transformed into matrix form based on horizon.
```
Comment (JasonZhou404) [inline docs/technical_documents/open_space_trajectory_optimizer_en.md:118]: based on horizon?

**Review by JasonZhou404 — CHANGES_REQUESTED** [review]

File: docs/technical_documents/hybrid_a_star_en.md (hunk 1 above)
Comment (storypku) [reply docs/technical_documents/hybrid_a_star_en.md:1]: in the open space ? on the open space. Please stick to the former.

File: docs/technical_documents/hybrid_a_star_en.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+![Diagram](images/opne_space_planner.png)
+
+1. Input:  The meaning of the input information is as follows: Vehicle current point(planned start point), goal point(planner end point),ROI_xy_boundary(the maximum and minimum boundary value of x and y), obstacles vertices vector(the corner position information). The function is follow:
+   ```  cpp
+        bool HybridAStar::Plan(double sx, double sy, double sphi, 
+                              double ex, double ey, double ephi,
+                              const std::vector<double>& XYbounds,
+                              const std::vector<std::vector<common::math::Vec2d>>& obstacles_vertices_vec,HybridAStartResult* result)
+   ```
+The input HybridAStar::Plan() is called by the open_space_trajectory_provider.cc, please refer(https://github.com/ApolloAuto/apollo/blob/master/modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_provider.cc)
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:19]: please refer [code]()

File: docs/technical_documents/hybrid_a_star_en.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
+Generate coarse trajectory on the open space 
+
+# Introduction
+The goal of this part is to generate the coarse trajectory in the open space. This part of the code consists of four parts: node3d，grid_search，reeds_shepp_path，hybrid_a_star. Hybrid_a_star the most important part of generating the coarse trajectory, it calls the grid_search and reeds_shepp_path.
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:4]: maybe: node3d，grid_search，reeds_shepp_path and hybrid_a_star.
Hybrid_a_star is the most important

File: docs/technical_documents/hybrid_a_star_en.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+1. Input:  The meaning of the input information is as follows: Vehicle current point(planned start point), goal point(planner end point),ROI_xy_boundary(the maximum and minimum boundary value of x and y), obstacles vertices vector(the corner position information). The function is follow:
+   ```  cpp
+        bool HybridAStar::Plan(double sx, double sy, double sphi, 
+                              double ex, double ey, double ephi,
+                              const std::vector<double>& XYbounds,
+                              const std::vector<std::vector<common::math::Vec2d>>& obstacles_vertices_vec,HybridAStartResult* result)
+   ```
+The input HybridAStar::Plan() is called by the open_space_trajectory_provider.cc, please refer(https://github.com/ApolloAuto/apollo/blob/master/modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_provider.cc)
+
+2. Process1: construct obstacles_linesegment_vector. The main method is to form a line segment from a single obstacle point in order; then, each obstacle line segment set is stored in obstacles_linesegment_vector that will be used to generate the DP map.
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:21]: "2. Process1", wrong format

File: docs/technical_documents/hybrid_a_star_en.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            for (size_t i = 0; i < vertices_num - 1; ++i) {
+            common::math::LineSegment2d line_segment = common::math::LineSegment2d(
+                obstacle_vertices[i], obstacle_vertices[i + 1]);
+            obstacle_linesegments.emplace_back(line_segment);
+            }
+            obstacles_linesegments_vec.emplace_back(obstacle_linesegments);
+        }
+        obstacles_linesegments_vec_ = std::move(obstacles_linesegments_vec);
+    ```
+3. Process2: construct the planned point same as Node3d,please refer (https://github.com/ApolloAuto/apollo/blob/master/modules/planning/open_space/coarse_trajectory_generator/node3d.h). The planned starting point and the ending point are constructed in the form of Node3d that will be save to open set and will be checked by the function of ValidityCheck().
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:37]: format of code reference
'space' after comma

File: docs/technical_documents/hybrid_a_star_en.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    ``` cpp
+        start_node_.reset(
+        new Node3d({sx}, {sy}, {sphi}, XYbounds_, planner_open_space_config_));
+        end_node_.reset(
+        new Node3d({ex}, {ey}, {ephi}, XYbounds_, planner_open_space_config_));
+    ```
+
+4. Process3: enter the main while loop to get the a set of nodes.
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:45]: the a set of ?

File: docs/technical_documents/hybrid_a_star_en.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+fisrt:Exit trajectory generation when open_pq_ is empty. The open_pq_ is a std::priority_queue type that the first element represents the order of nodes in the open table, the second element represents the cost of nodes which storage in descending order.
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:46]: maybe: first: exit

File: docs/technical_documents/hybrid_a_star_en.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+4. Process3: enter the main while loop to get the a set of nodes.
+fisrt:Exit trajectory generation when open_pq_ is empty. The open_pq_ is a std::priority_queue type that the first element represents the order of nodes in the open table, the second element represents the cost of nodes which storage in descending order. 
+
+second: Use AnalyticExpansion() function to determine whether there is a collision free trajectory that based on the reeds_shepp_path from current point to target end point. if it exists, exit the while loop search.
+    ``` cpp
+        if (AnalyticExpansion(current_node)) {
+            break;
+        }
+    ```
+third: Store the current point in the close table and According to the number of samples set for the next node, the cycle mode is carried out. generate the next node by Next_node_generator() and use ValidityCheck() to detect this node.
```
Comment (spicyog) [inline docs/technical_documents/hybrid_a_star_en.md:54]: according

**Comment by spicyog** [comment]
Some suggestion:
1. Use the secondary title to refine your document format.
2. Code style ```thefunction()``` could be used if it's a function.

**Pull request closed without merging**